rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,394 @@
+
+### Source of Truth
+
+TensorFlow uses a Google-internal code repository as its source of truth. Every PR
+submitted though GitHub is converted to a Google-internal change first,
+submitted through the internal system, and then copied to GitHub as commits.
+At the same time, PR is marked as merged with the corresponding commit hash.
+
+Likewise, issue tracking and code review takes place through Google-internal tools.
+
+For Keras, since we are trying to promote community engagement, we hope to use 
+GitHub as source of truth. This will have the following implications:
+
+* We expect the majority of the code development/contribution from GitHub
+and the dev tools / tests / scripts should focus on the GitHub development use
+case. See below for more details.
+* Keras CI/presubmit build for the GitHub repo should target a stable PIP 
+version of tensorflow package as dependency. It could either be (preferably in
+this order):
+  * a stable version
+  * a release candidate version
+  * a `tf-nightly` with explicit version.
+Using a nightly version for testing should be motivated by the usage of a API
+feature not present in the stable or pre-release version.
+Depend on a floating `tf-nightly` could cause CI build to be instable, which has
+been observed in other repository 
+[like tf-addons](https://github.com/tensorflow/addons/pull/912).
+* The Keras code will be mirrored to a Google-internal code repository via
+Google-internal tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
+* The CI build for the repository on GitHub might break when it points to a
+new version of `tf-nightly`, if certain behavior has been changed and wasn't
+caught by unit tests. We have observed a few similar cases with
+[tf/addons](https://github.com/tensorflow/addons).
+We hope this can be reduced by stronger unit test coverage by Google internel
+systems, when both TF and Keras code are tested at HEAD.
+* pip package management. Keras will now follow the `tf-estimator` approach. 
+"pip install tensorflow" should also install Keras (from PyPI) as well.
+There are more details for the pip package in the
+[Improved pip package structure](https://github.com/tensorflow/community/pull/182) RFC.
+
+### Dependency Cleanup
+
+As the high-level API of TensorFlow, Keras should have a direct dependency on
+TF low-level APIs, but not the other way around. Unfortunately, there is some existing reverse 
+logic in the TF code that relies on Keras, which we should update/remove 
+when we split the repository.
+
+The current usage of Keras from TensorFlow are:
+* Unit tests, which should be converted to integration tests, or port the tests
+to Keras repository.
+* `feature_column`, which uses Keras base layer and model.
```

Comment:
These dependencies are very concerning.

Does this mean we need a third package (tf, tf-keras, and tf-stuff-that-depends-on-keras)? Where would the code for this third package live? How can we test its features?

Or are splitting all the things which depend on keras despite not being historically a part of the keras API into the keras repo?

Reply:
I think the third package u mentioned can be avoided by the LazyLoader. For any tf code that relies on Keras, it will need keras PIP package to be installed to work. Since PIP install TF will also install Keras package, so the keras code should always be there unless user manually changed it.

Reply:
Had some email discussion with TF API owner. So far the API owners are OK if the TF code that relies on Keras is V1 and only use public Keras API (eg layer and model). For feature column, we might want to move them to estimator package, and let estimator depend on keras.